### PR TITLE
chore(test-data): create `ProductDiscount` model and associated draft

### PR DESCRIPTION
**Pull request by jaikamat**
## Summary
This PR creates the [ProductDiscount](https://docs.commercetools.com/api/projects/productDiscounts#productdiscount) entity model.

## Description
Nothing out of the ordinary here. Additional comments inline

### CODE REVIEW

File: models/product-discount/src/generator.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+} from '@commercetools-test-data/commons';
+import { createRelatedDates } from '@commercetools-test-data/utils';
+import * as ProductDiscountValueAbsolute from '@commercetools-test-data/product-discount-value-absolute';
+import * as ProductDiscountValueRelative from '@commercetools-test-data/product-discount-value-relative';
+import * as ProductDiscountValueExternal from '@commercetools-test-data/product-discount-value-external';
+
+// https://docs.commercetools.com/api/projects/productDiscounts#productdiscount
+
+// eslint-disable-next-line
+const [getValidFrom, _, getValidUntil] = createRelatedDates();
```
Comment (jaikamat) [inline models/product-discount/src/generator.ts:15]: the third element of this tuple is a "future date", which I thought would be useful for the `validUntil` property. To access it, I had to squelch the eslint error when setting the middle element to unused.

File: models/product-discount/src/generator.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+      f.helpers.arrayElement([
+        ProductDiscountValueAbsolute.random(),
+        ProductDiscountValueRelative.random(),
+        ProductDiscountValueExternal.random(),
+      ])
+    ),
+    predicate: '1=1',
+    sortOrder: fake((f) => String(Math.random())),
+    isActive: fake((f) => f.datatype.boolean()),
+    references: [],
```
Comment (jaikamat) [inline models/product-discount/src/generator.ts:35]: We can resolve these references in presets. As they need to relate to items referenced in the predicate, doing this here might cause some unnecessary complexity.

File: yarn.lock [hunk 3]
```diff
@@ -2022,6 +2022,23 @@ __metadata:
...
+  version: 0.0.0-use.local
+  resolution: "@commercetools-test-data/product-discount@workspace:models/product-discount"
+  dependencies:
+    "@babel/runtime": ^7.17.9
+    "@babel/runtime-corejs3": ^7.17.9
+    "@commercetools-test-data/commons": 4.6.0
+    "@commercetools-test-data/core": 4.6.0
+    "@commercetools-test-data/product-discount-value-absolute": 4.6.0
+    "@commercetools-test-data/product-discount-value-external": 4.6.0
+    "@commercetools-test-data/product-discount-value-relative": 4.6.0
```
Comment (jaikamat) [inline yarn.lock:2035]: @emmenko What're your thoughts on these discount value models? Should they all be part of one package?

I broke them out for obviousness but was wondering whether union type models should be grouped in the same directory.

**Review by stephsprinkle — APPROVED** [review]

File: models/product-discount/src/generator.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    description: fake(() => LocalizedString.random()),
+    value: fake((f) =>
+      f.helpers.arrayElement([
+        ProductDiscountValueAbsolute.random(),
+        ProductDiscountValueRelative.random(),
+        ProductDiscountValueExternal.random(),
+      ])
+    ),
+    predicate: '1=1',
+    sortOrder: fake((f) => String(Math.random())),
```
Comment (stephsprinkle) [inline models/product-discount/src/generator.ts:33]: This need to be between 0 and 1

File: models/product-discount/src/product-discount-draft/generator.ts [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    description: fake(() => LocalizedString.random()),
+    value: fake((f) =>
+      f.helpers.arrayElement([
+        ProductDiscountValueAbsolute.ProductDiscountValueAbsoluteDraft.random(),
+        ProductDiscountValueRelative.ProductDiscountValueRelativeDraft.random(),
+        ProductDiscountValueExternal.ProductDiscountValueExternalDraft.random(),
+      ])
+    ),
+    predicate: '1=1',
+    sortOrder: fake((f) => String(Math.random())),
```
Comment (stephsprinkle) [inline models/product-discount/src/product-discount-draft/generator.ts:27]: Same here

File: models/product-discount/src/generator.ts (hunk 1 above)
Comment (tylermorrisford) [reply models/product-discount/src/generator.ts:15]: 💪

File: models/product-discount/src/product-discount-draft/transformers.ts [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    'graphql',
+    {
+      buildFields: ['value', 'name', 'description'],
+      addFields: ({ fields }) => ({
+        nameAllLocales: LocalizedString.toLocalizedField(fields.name),
+        descriptionAllLocales: LocalizedString.toLocalizedField(
+          fields.description
+        ),
+        __typename: 'ProductDiscountDraft',
+      }),
```
Comment (stephsprinkle) [inline models/product-discount/src/product-discount-draft/transformers.ts:26]: I don't see these fields in the GraphQL docs

image

Comment (jaikamat) [reply models/product-discount/src/product-discount-draft/transformers.ts:26]: Whoooooopsie! Thanks for checking so thoroughly!

File: models/product-discount/src/builder.spec.ts [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+        sortOrder: expect.any(String),
+        isActive: expect.any(Boolean),
+        references: expect.arrayContaining([]),
+        validFrom: expect.any(String),
+        validUntil: expect.any(String),
+        createdAt: expect.any(String),
+        createdBy: expect.objectContaining({
+          customerRef: expect.objectContaining({ typeId: 'customer' }),
+          userRef: expect.objectContaining({ typeId: 'user' }),
+          __typename: expect.any(String),
```
[the remark this answers is not in the document]
Comment (jaikamat) [reply models/product-discount/src/builder.spec.ts:110]: Yep! That was me being lazy 😆

File: models/product-discount/src/generator.ts (hunk 4 above)
Comment (jaikamat) [reply models/product-discount/src/generator.ts:33]: Gahhh `Math.random()` could be zero, couldn't it

File: models/product-discount/src/builder.spec.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (tylermorrisford) [reply models/product-discount/src/builder.spec.ts:110]: 🎵 you are forgivennnnn 🎶

**Review by tylermorrisford — APPROVED** [review]

File: models/product-discount/src/product-discount-draft/generator.ts (hunk 5 above)
Comment (jaikamat) [reply models/product-discount/src/product-discount-draft/generator.ts]: Thanks! See b6c9265df2712639c9c23d548783ff7fb4a84fec

File: models/product-discount/src/builder.spec.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (jaikamat) [reply models/product-discount/src/builder.spec.ts]: 😅 See 0227e44b70becfc06112cfc0f7b70330b9de67bc

File: models/product-discount/src/product-discount-draft/transformers.ts (hunk 6 above)
Comment (jaikamat) [reply models/product-discount/src/product-discount-draft/transformers.ts]: Resolved in 7dde2eedf72087d87efe2186104a3a1b12eab5fb

File: models/product-discount/package.json [hunk 8]
```diff
@@ -0,0 +1,31 @@
+{
+  "name": "@commercetools-test-data/product-discount",
+  "version": "4.6.0",
```
[the remark this answers is not in the document]
Comment (jaikamat) [reply models/product-discount/package.json]: Nice! See de727c67be42d3fbe37a2727c6df379ab0024dd5

File: models/product-discount/src/types.ts [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+  __typename: 'ProductDiscount';
+  createdBy: TClientLoggingGraphql;
+  lastModifiedBy: TClientLoggingGraphql;
+  nameAllLocales?: TLocalizedStringGraphql | null;
+  descriptionAllLocales?: TLocalizedStringGraphql | null;
+};
+export type TProductDiscountDraftGraphql = TProductDiscountDraft & {
+  __typename: 'ProductDiscountDraft';
+  nameAllLocales?: TLocalizedStringGraphql | null;
+  descriptionAllLocales?: TLocalizedStringGraphql | null;
```
Comment (stephsprinkle) [inline models/product-discount/src/types.ts:24]: Don't need these

Comment (jaikamat) [reply models/product-discount/src/types.ts]: 🦅 5d65eb548464ed6895f9707020cfec95231429ed

**Review by stephsprinkle — APPROVED** [review]
This looks good, I'm really done reviewing now!

**Pull request merged**